lib/PersonalityProvider.jsm

```diff
@@ -0,0 +1,197 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {PersistentCache} = ChromeUtils.import("resource://activity-stream/lib/PersistentCache.jsm", {});
+const {RemoteSettings} = ChromeUtils.import("resource://services-settings/remote-settings.js", {});
+
+const {NaiveBayesTextTagger} = ChromeUtils.import("resource://activity-stream/lib/NaiveBayesTextTagger.jsm", {});
+const {NmfTextTagger} = ChromeUtils.import("resource://activity-stream/lib/NmfTextTagger.jsm", {});
+const {RecipeExecutor} = ChromeUtils.import("resource://activity-stream/lib/RecipeExecutor.jsm", {});
+
+ChromeUtils.defineModuleGetter(this, "NewTabUtils",
+  "resource://gre/modules/NewTabUtils.jsm");
+
+const STORE_UPDATE_TIME = 24 * 60 * 60 * 1000; // 24 hours
+
+/**
+ * V2 provider builds and ranks an interest profile (also called an “interest vector”) off the browse history.
+ * This allows Firefox to classify pages into topics, by examining the text found on the page.
+ * It does this by looking at the history text content, title, and description.
+ */
+this.PersonalityProvider = class PersonalityProvider {
+  constructor(
+    timeSegments,
+    parameterSets,
+    maxHistoryQueryResults,
+    version,
+    scores,
+    modelKeys) {
+    this.modelKeys = modelKeys;
+    this.timeSegments = timeSegments;
+    this.parameterSets = parameterSets;
+    this.maxHistoryQueryResults = maxHistoryQueryResults;
+    this.version = version;
+    this.scores = scores;
+    this.store = new PersistentCache("personality-provider", true);
+  }
+
+  async init() {
+    this.interestConfig = await this.getRecipe();
+    this.recipeExecutor = await this.generateRecipeExecutor();
+    this.interestVector = await this.store.get("interest-vector");
+
+    // Fetch a new one if none exists or every set update time.
+    if (!this.interestVector ||
+      (Date.now() - this.interestVector.lastUpdate) >= STORE_UPDATE_TIME) {
+      this.interestVector = await this.createInterestVector();
+      this.interestVector.lastUpdate = Date.now();
+      this.store.set("interest-vector", this.interestVector);
+    }
+    this.initialized = true;
+  }
+
+  getRemoteSettings(name) {
```

Comment:
nit: the function name sounds more like a getter for the RemoteSettings property, but it's actually a lookup function on RemoteSettings. So let's rename it (e.g. `getFromRemoteSettings`) to avoid ambiguity.

Reply:
Another question about the RemoteSettings, are you going to use multiple Remote Settings buckets for this feature? Because the `name` parameter specifies the bucket name in `RemoteSettings(name).get()`. I believe the idiomatic way of using Remote Settings is that creating multiple records in a single bucket for this feature, and use `filters` to look up specific records. For instance,

Bucket name: "Pocket-Personalization"
```js
{
  "data": [
     {"name": "personality-provider-models", ...},
     {"name": "other-models", ...},
  ]
}
``` 

Then you can query it with 
```js
RemoteSettings("Pocket-Personalization").get({filters: {name: "personality-provider-models"}});
```

Reply:
I currently have two collections, 1 is "personality-provider-recipe" which there is really only 1 record for. The other is personality-provider-models which has many records.

The only reason I kept them separate is it looked like collections were expected to have similar data structures. personality-provider-recipe was something different than personality-provider-models, so it didn't look correct to add personality-provider-recipe and personality-provider-models to the same collection. I'm a bit confused though.

Reply:
Okay, looks like having two separate buckets is expected here.